### ISSUE TITLE
Illustrate features of Pony with Code Samples

### ISSUE THREAD

**Issue by mfelsche** [ponylang/ponylang-website#292]
In this hackernews conversation: https://news.ycombinator.com/item?id=17195580 , lots of people demanded code examples on the website. They couldn't find any, even after clicking around.

I think it makes sense to give shallow first impression with some code example that highlights some of the strengths of pony and to provide some more examples on a separate page, centered around how to do common things in pony, people might be used to from other languages.

**Comment by SeanTAllen** [comment]
I disagree.

**Comment by SeanTAllen** [comment]
@mfelsche and I chatted a bit. I'm quite good with approaching this as "impression should someone have after a quick visit to the homepage" and working from there to figure out what in terms of text, code, images, videos etc should be there.  

The key for me is having a goal besides "put some code on there". We should have a defined goal and figure out how to meet that goal via whatever means (including code).

**Comment by mfelsche** [comment]
The main goal i would like to pursue is to have the visitor of the website, especially newcomers to pony, be able to get a good first impression of the language and if they want lead them to other sites for further information (like the tutorial or the stdlib docs). 
Seeing the actual language as code example is not necessarily part of such a first expression, but it is something people are interested in, in general. And this desire is understandable as we are exhibiting a programming language here. The main interface to the language is writing it, so people have a legitimate interest in seeing what they might have to write to make certain things work or leverage certain properties of the Pony runtime... 

So, to clarify, i don't want code examples for the sake of having them, but for the following reasons:
   
  * Syntax is an integral part of the language, although not the most interesting part of Pony, i have to admit. (Furthermore the pony syntax is nothing to be ashamed of. I do like it a lot.) Hence, a visitor who would like to know more about the language should also be able to find out, what it looks like.
  * Code examples are a great way of illustrating features of a programming language. The website is full of descriptions of the features of Pony, these would be more insightful if illustrated with code examples. They also serve the purpose of teaching people can leverage or work with certain properties of Pony. (Say e.g. the property that messages sent to actors are only pointers being moved through mesage queues, finally passed into some actor behaviour. Seeing how simple this is by the help of a code sample might make people appreciate the language as much as i do.)

**Comment by plietar** [comment]
A central aspect that the front page should show off is actors. Most readers coming from other languages will not have used them before, and even people coming from eg Erlang have used a very different style. This is IMO what really differentiates Pony from other languages (maybe I'm biased, coming from Rust/C++).

But just having text explaining what actors are is not going to grab attention.

If we had a simple example (like two actors, just doing ping pong) that demonstrates how easy and intuitive actors are in Pony I think it would help a lot.

**Comment by SeanTAllen** [comment]
I agree with @plietar re: actors. I think something that shows actors is certainly part of it. Hello world while there is more than 1 actor is very non-obvious and wouldn't be a good idea in my mind. Ping/Pong might be good.

Ref caps are an integral part of Pony as well but, do we show them that early on? I think not. Or rather, we don't highlight them.

I think something like Ping/Pong with an explanation of what is going on could be good.

Showing code for the sake of syntax is in my mind, unimportant. If you are coming for to see the syntax, you aren't getting over the hump of reference capabilities. That's the wrong audience for us.

So, if you want things to highlight I think it should be:

safe concurrency via actors. that's the hard problem that pony is solving. its what needs to be emphasized. I'm not against code samples but they should be specifically to demonstrate what makes Pony different.

Its a serious investment in time to learn Pony. People need to be presented with a case for why they should do that.

**Comment by SeanTAllen** [comment]
One additional note, I'm worried that Ping/Pong wouldnt end up demonstrating enough. We might need something a bit more interesting. Find a good example that is accesible in terms of problem domain yet also demonstrates what we want could be hard. 

It might require more than 1 example but a single example would be best in my mind. Perhaps something from `examples` in the repo.

**Comment by thomastay** [comment]
Reviving this old issue as I've been thinking about this too. I also disagree that code samples are that important, especially since syntax isn't Pony's strength. 

I think it's most insightful to take a leaf from other languages' websites:
https://www.rust-lang.org/ 

- Has a splash logo, the language name in HUGE FONT, a tagline "A language empowering everyone
to build reliable and efficient software.", and a "Get Started" button. All things which are immediate in drawing attention. Importantly, users **self-classify** : those who want to use Rust will get started, users who aren't conviced will read on.
- On that note, the Get Started page takes them immediately to the install page. There is no choice in this matter, which is good in guiding users. Remember: this target audience is users who already want to use the language.
- For users who aren't convinced, scrolling down the homepage brings you to "Why Rust?", with three taglines: 

1. Performance - Rust is blazingly fast and memory-efficient: with no runtime or garbage collector, it can power performance-critical services, run on embedded devices, and easily integrate with other languages.
1. Reliability - Rust’s rich type system and ownership model guarantee memory-safety and thread-safety — enabling you to eliminate many classes of bugs at compile-time.
1. Productivity - Rust has great documentation, a friendly compiler with useful error messages, and top-notch tooling — an integrated package manager and build tool, smart multi-editor support with auto-completion and type inspections, an auto-formatter, and more.

- The Homepage then follows with 4 classes of apps that can be built in Rust, along with 4 links for the unconvinced user to **self-classify** themselves, and read the guide *that is tailored* for that particular group. When you click on 1 of the 4 links, **now** only do you see the code samples. 
- You see a Testimonials section, where there are big splash quotes from prominent users who've used Rust.
- Then, a "Get Involved" section. Personally, I think that section is unhelpful for new users and is mainly for existing users who want to deepen their knowledge.
- Laslty, a "Thanks" section, linking to their backer page.

**Comment by damon-kwok** [comment]
@mfelsche Is this what you want?
![nim](https://user-images.githubusercontent.com/1702133/84978242-f7c2c380-b15e-11ea-93a0-490b5bf7b188.png)
![crystal](https://user-images.githubusercontent.com/1702133/84978246-fa251d80-b15e-11ea-9c9e-3ab4d17f515c.png)
![elixir](https://user-images.githubusercontent.com/1702133/84978248-fabdb400-b15e-11ea-9863-d1b95ce8654f.png)

@SeanTAllen We can open a discussion topic in zulip and collect the most representative code of pony. It must have the characteristics of pony and the most spiritual representative. I think it's good if it is up to expectations. When I first logged in to ponylang.io, I was actually very eager to see this.

**Comment by SeanTAllen** [comment]
@damon-kwok I will argue against all those as a waste of space.

**Comment by fenginsc** [comment]
It's now 2025 and I think the homepage should be redesigned to look more aesthetically pleasing and modern, it's just too minimal at the moment!

**Comment by SeanTAllen** [comment]
We added a link to examples on the homepage years ago, so we are closing this. The original ticket has been addressed.

<img width="1359" height="994" alt="Image" src="https://github.com/user-attachments/assets/12ecde95-2d82-45a9-bce4-7c6df2ea033e" />
